draft-ietf-httpbis-message-signatures.md

```diff
@@ -1283,22 +1291,161 @@ The table below contains the initial contents of the HTTP Signature Specialty Co
+
+To counteract these kinds of attacks, it's first important for the signer to cover sufficient portions of the message to differentiate it from other messages. In addition, the signature can use the `nonce` signature parameter to provide a per-message unique value to allow the verifier to detect replay of the signature itself if a nonce value is repeated. Furthermore, the signer can provide a timestamp for when the signature was created and a time at which the signer considers the signature to be invalid, limiting the utility of a captured signature value.
+
+If a verifier wants to trigger a new signature from a signer, it can send the `Accept-Signature` header field with a new `nonce` parameter. An attacker that is simply replaying a signature would not be able to generate a new signature with the chosen nonce value.
+
+## Insufficient Coverage {#security-coverage}
+
+Any portions of the message not covered by the signature are susceptible to modification by an attacker without affecting the signature. An attacker can take advantage of this by introducing a header field or other message component that will change the processing of the message but will not be covered by the signature. Such an altered message would still pass signature verification, but when the verifier processes the message as a whole, the unsigned content injected by the attacker would subvert the trust conveyed by the valid signature and change the outcome of processing the message.
+
+To combat this, an application of this specification should require as much of the message as possible to be signed, within the limits of the application and deployment. The verifier should only trust message components that have been signed. Verifiers could also strip out any sensitive unsigned portions of the message before processing of the message continues.
+
+## Cryptography and Signature Collision {#security-collision}
+
+The HTTP Message Signatures specification does not define any of its own cryptographic primitives, and instead relies on other specifications to define such elements. If the signature algorithm or key used to process the signature input string is vulnerable to any attacks, the resulting signature will also be susceptible to these same attacks.
+
+A common attack against signature systems is to force a signature collision, where the same signature value successfully verifies against multiple different inputs. Since this specification relies on reconstruction of the input string based on an HTTP message, and the list of components signed is fixed in the signature, it is difficult but not impossible for an attacker to effect such a collision. An attacker would need to manipulate the HTTP message and its covered message components in order to make the collision effective.
```

Comment:
These security considerations make me want to reframe signature verification as "verify a signature against your requirements" rather than just "verify a signature against its input". E.g., verification steps are:
1. Identify the set of message components required/expected/allowed to be signed, keys and algs permitted, etc.
2. Verify that the signature input matches requirements.
3. Verify that the corresponding signature matches signature input and message. (i.e., existing verification instructions)

This highlights the importance of identifying requirements and testing against those first and foremost - a "valid" signature that doesn't cover what's needed is irrelevant - and emphasizes this as a critical part of the verifier's job. As it is, this is tacked on at the end of the verification process. I want to encourage library interfaces that look like `verify(message, required_components)` not `verify(message)`.

Reply:
I agree with that in principle, and I get where it's coming from. We've got some precedent in the world of JOSE implementations where you send not only the JWS but also the required header and (sometimes) body claims for a JWT. My only concern is that there is a separate definitive step of doing the signature verification calculation, and we don't want that to get lost in the overall discussion. How can we frame it so that we mostly talk about the larger piece but still can call out the smaller piece when necessary? Namely when we define how the algorithms are applied.